src/main/scala/org/broadinstitute/dsde/workbench/sam/service/PolicyEvaluatorService.scala

```diff
@@ -130,6 +120,40 @@ class PolicyEvaluatorService(
       publicPolicies <- accessPolicyDAO.listPublicAccessPolicies(resourceTypeName, samRequestContext)
     } yield results.flatten ++ publicPolicies.map(p => UserPolicyResponse(p.resourceId, p.accessPolicyName, Set.empty, Set.empty, public = true))
 
+  def listUserResources(resourceTypeName: ResourceTypeName, userId: WorkbenchUserId, samRequestContext: SamRequestContext): IO[Iterable[UserResourcesResponse]] =
+    for {
+      rt <- getResourceType(resourceTypeName)
+      isConstrained = rt.isAuthDomainConstrainable
+      resourcesWithAccess <- accessPolicyDAO.listUserResourcesWithRolesAndActions(resourceTypeName, userId, samRequestContext)
+      rids = resourcesWithAccess.map(_.resourceId).toSet
+
+      resources <- if (isConstrained) accessPolicyDAO.listResourcesWithAuthdomains(resourceTypeName, rids, samRequestContext)
+      else IO.pure(Set.empty)
+      authDomainMap = resources.map(x => x.resourceId -> x.authDomain).toMap
+
+      userManagedGroups <- if (isConstrained) listUserManagedGroups(userId, samRequestContext) else IO.pure(Set.empty[WorkbenchGroupName])
+
+      results = resourcesWithAccess.map { resourceWithAccess =>
+        if (isConstrained) {
```

Comment:
Could this method be rewritten and simplified by making the `isConstrained` check earlier and only one time, so as to isolate all the logic used in the constrained path from the seemingly simple unconstrained path.